File: lib/src/resume-ci.ts
#!/usr/bin/env bun

import { parseArgs } from "node:util"
import { mkdirSync, readdirSync, statSync } from "node:fs"
import { join, relative, resolve } from "node:path"
import { parse as parseYaml } from "yaml"
import { z } from "zod"
import { resumeSchema, role, education, type Resume } from "./schema.ts"
import { rich, periodStr, extractDomain, extractUsername } from "./utils.ts"

const ROOT      = resolve(import.meta.dir, "../..")
const DATA_DIR  = join(ROOT, "resumes")
const BUILD_DIR = join(ROOT, "build")
const FONT_DIR  = join(ROOT, "lib", "bin", "fonts")
const TEMPLATE  = join(ROOT, "templates", "default.typ")


const SECTION_TITLES = {
  summary: "Professional Summary", experience: "Experience", projects: "Projects",
  certifications: "Certifications", education: "Education", skills: "Technical Skills",
}

type ContactItem = { icon: string; solid: boolean; href: string; text: string }

function buildContacts(p: Resume["personal"]): ContactItem[] {
  return [
    { icon: "envelope",  solid: true,  href: `mailto:${p.email}`, text: p.email },
    p.phone        ? { icon: "phone",    solid: true,  href: "", text: p.phone } : null,
    p.location     ? { icon: "",         solid: false, href: "", text: p.location } : null,
    p.linkedin_url ? { icon: "linkedin", solid: false, href: p.linkedin_url, text: extractUsername(p.linkedin_url) } : null,
    p.github_url   ? { icon: "github",   solid: false, href: p.github_url,   text: extractUsername(p.github_url) } : null,
  ].filter((x): x is ContactItem => x !== null)
}

function buildRole(item: z.infer<typeof role>) {
  const u = item.url ?? ""
  return { company: rich(item.company), period: periodStr(item.period), role: rich(item.role), url: u, domain: u ? extractDomain(u) : "", bullets: item.bullets.map(rich) }
}

function buildContext(data: Resume): Record<string, unknown> {
  return {
    font:           data.font,
    section_titles: { ...SECTION_TITLES, ...(data.section_titles ?? {}) },
    personal:       { name: rich(data.personal.name), title: rich(data.personal.title) },
    contact:        buildContacts(data.personal),
    summary:        rich(data.summary ?? ""),
    experience:     data.experience.map(buildRole),
    projects:       data.projects.map(buildRole),
    certifications: data.certifications.map(rich),
    education:      data.education.map(item => ({
      institution: rich(item.institution), period: periodStr(item.period),
      degree: rich(item.degree), location: rich(item.location),
    })),
    skills:         data.skills.map(item => ({ label: rich(item.label), items: rich(item.items) })),
    output_filename: data.output_filename,
  }
}


class Builder {
  private constructor(
    private readonly templatePath: string,
    private readonly outputDir: string,
    private readonly paths: string[],
  ) {}

  static create(templatePath: string, outputDir: string, explicit: string[] | null) {
    try { statSync(templatePath) } catch {
      throw new Error(`Template not found: ${templatePath}`)
    }
    return new Builder(templatePath, outputDir, Builder.resolvePaths(explicit))
  }

  async buildAll(): Promise<boolean> {
    let failed = false
    for (const path of this.paths) {
      try { console.log(`PDF: ${await this.compile(path)}`) }
      catch (err) { console.error(`FAILED ${path}: ${(err as Error).message}`); failed = true }
    }
    return failed
  }

  async watch(): Promise<void> {
    const watched = [...this.paths, this.templatePath]
    let last: string | null = null

    process.on("SIGINT", () => { console.log("\nStopped watching."); process.exit(0) })
    console.log("Watching for changes. Press Ctrl+C to stop.")

    while (true) {
      const snapshot = JSON.stringify(watched.map(f => { try { return statSync(f).mtimeMs } catch { return 0 } }))
      if (snapshot !== last) { await this.buildAll(); last = snapshot }
      await Bun.sleep(500)
    }
  }

  private async compile(resumePath: string): Promise<string> {
    const raw = parseYaml(await Bun.file(resumePath).text())

    const result = resumeSchema.safeParse(raw)
    if (!result.success) {
      const issue = result.error.issues[0]
      const path = issue.path.join(".") || "resume"
      throw new Error(`${path}: ${issue.message}`)
    }

    const ctx = buildContext(result.data)
    mkdirSync(this.outputDir, { recursive: true })
    const pdf = join(this.outputDir, `${ctx.output_filename}.pdf`)

    const proc = Bun.spawn(
      [Builder.findTypst(), "compile", "--root", ROOT, "--font-path", FONT_DIR,
       "--input", `data=${JSON.stringify(ctx)}`, relative(ROOT, this.templatePath), pdf],
      { cwd: ROOT, stderr: "inherit" },
    )
    if (await proc.exited !== 0) throw new Error("typst compile failed")
    return pdf
  }

  private static resolvePaths(explicit: string[] | null): string[] {
    if (explicit) return explicit
    const files = readdirSync(DATA_DIR)
      .filter(f => f.endsWith(".yml") && !f.startsWith("."))
      .sort()
      .map(f => join(DATA_DIR, f))
    if (!files.length) { console.error("No resume YAML files found."); process.exit(0) }
    return files
  }

  private static findTypst(): string {
    for (const name of ["typst", "typst.exe"]) {
      const path = join(ROOT, "lib", "bin", name)
      try { statSync(path); return path } catch {}
    }
    const found = Bun.which("typst")
    if (found) return found
    throw new Error("typst not found. Run make setup")
  }
}


const { values, positionals } = parseArgs({
  args: Bun.argv.slice(2),
  options: {
    template:     { type: "string",  default: TEMPLATE },
    "output-dir": { type: "string",  default: BUILD_DIR },
    watch:        { type: "boolean", default: false },
  },
  allowPositionals: true,
})

const builder = Builder.create(
  resolve(values.template as string),
  values["output-dir"] as string,
  positionals.length ? positionals.map(p => resolve(p)) : null,
)

if (values.watch) {
  await builder.watch()
} else {
  process.exit(await builder.buildAll() ? 1 : 0)
}

File: lib/resume-ci.py
#!/usr/bin/env python3

import argparse
import json
import re
import shutil
import subprocess
import sys
import time
from pathlib import Path
from urllib.parse import urlparse

import yaml
from jsonschema import Draft202012Validator, FormatChecker

ROOT = Path(__file__).resolve().parent.parent
DATA_DIR = ROOT / "resumes"
BUILD_DIR = ROOT / "build"
TEMPLATE = ROOT / "templates" / "default.typ"
FONT_DIR = ROOT / "bin" / "fonts"

TITLES = {
    "summary": "Professional Summary",
    "experience": "Experience",
    "projects": "Projects",
    "certifications": "Certifications",
    "education": "Education",
    "skills": "Technical Skills",
}

RICH_RE = re.compile(r"\*\*(.+?)\*\*|_(.+?)_")
SCHEMA = Draft202012Validator(
    json.loads((ROOT / "lib" / "resume.schema.json").read_text()),
    format_checker=FormatChecker(),
)


def parse_args():
    parser = argparse.ArgumentParser(description="Build resume PDFs from YAML files.")
    parser.add_argument("resumes", nargs="*", type=Path)
    parser.add_argument("--template", type=Path, default=TEMPLATE)
    parser.add_argument("--output-dir", type=Path, default=BUILD_DIR)
    parser.add_argument("--watch", action="store_true")
    return parser.parse_args()


def paths(explicit):
    found = explicit or sorted(DATA_DIR.glob("*.yml"))
    if not found:
        print("No resume YAML files found.", file=sys.stderr)
        sys.exit(0)
    return found


def load(path):
    data = yaml.safe_load(path.read_text())
    if not isinstance(data, dict):
        raise ValueError(f"{path} must contain a YAML mapping")

    errors = sorted(SCHEMA.iter_errors(data), key=lambda error: list(error.path))
    if errors:
        error = errors[0]
        location = ".".join(str(part) for part in error.absolute_path) or "resume"
        raise ValueError(f"{location}: {error.message}")
    return data


def text(data, key, default=""):
    return (data.get(key) or default).strip()


def rich(value):
    parts = []
    pos = 0
    for match in RICH_RE.finditer(value):
        if match.start() > pos:
            parts.append({"text": value[pos:match.start()], "style": ""})
        parts.append({"text": match.group(1) or match.group(2), "style": "strong" if match.group(1) else "emph"})
        pos = match.end()
    if pos < len(value):
        parts.append({"text": value[pos:], "style": ""})
    return parts


def period(item):
    value = item.get("period") or {}
    start = text(value, "from")
    end = text(value, "to")
    return f"{start} -- {end}" if start and end else start or end


def domain(url):
    host = urlparse(url).hostname or ""
    parts = host.split(".")
    keep = 3 if re.search(r"\.(com|org|net|gov|edu)\.[a-z]{2}$", host) else 2
    return ".".join(parts[-keep:]) if len(parts) > keep else host


def username(url):
    path = urlparse(url).path.strip("/")
    return path.split("/")[-1] if path else re.sub(r"^https?://", "", url)


def contacts(personal):
    email = text(personal, "email")
    items = [{"icon": "envelope", "solid": True, "href": f"mailto:{email}", "text": email}]
    for key, icon, solid in (("phone", "phone", True), ("location", "", False)):
        if value := text(personal, key):
            items.append({"icon": icon, "solid": solid, "href": "", "text": value})
    for key, icon in (("linkedin_url", "linkedin"), ("github_url", "github")):
        if url := text(personal, key):
            items.append({"icon": icon, "solid": False, "href": url, "text": username(url)})
    return items


def role(item):
    url = text(item, "url")
    return {
        "company": rich(text(item, "company")),
        "period": period(item),
        "role": rich(text(item, "role")),
        "url": url,
        "domain": domain(url) if url else "",
        "bullets": [rich(bullet) for bullet in item.get("bullets", [])],
    }


def education(item):
    return {
        "institution": rich(text(item, "institution")),
        "period": period(item),
        "degree": rich(text(item, "degree")),
        "location": rich(text(item, "location")),
    }


def context(data):
    personal = data["personal"]
    return {
        "personal": {"name": rich(text(personal, "name")), "title": rich(text(personal, "title"))},
        "contact": contacts(personal),
        "summary": rich(text(data, "summary")),
        "font": text(data, "font", "New Computer Modern"),
        "section_titles": {**TITLES, **(data.get("section_titles") or {})},
        "experience": [role(item) for item in data.get("experience", [])],
        "projects": [role(item) for item in data.get("projects", [])],
        "certifications": [rich(item) for item in data.get("certifications", [])],
        "education": [education(item) for item in data.get("education", [])],
        "skills": [{"label": rich(text(item, "label")), "items": rich(text(item, "items"))} for item in data.get("skills", [])],
        "output_filename": data["output_filename"],
    }


def typst():
    for name in ("typst", "typst.exe"):
        path = ROOT / "bin" / name
        if path.exists():
            return str(path)
    if path := shutil.which("typst"):
        return path
    raise RuntimeError("typst not found. Run lib/setup.sh")


def root_path(path):
    return path.resolve().relative_to(ROOT).as_posix()


def build(path, template, output_dir):
    data = context(load(path))
    output_dir.mkdir(parents=True, exist_ok=True)
    pdf = output_dir / f"{data['output_filename']}.pdf"
    payload = json.dumps(data, ensure_ascii=False, separators=(",", ":"))
    subprocess.run([
        typst(), "compile",
        "--root", str(ROOT),
        "--font-path", str(FONT_DIR),
        "--input", f"data={payload}",
        root_path(template), str(pdf),
    ], check=True)
    return pdf


def build_all(resumes, template, output_dir):
    failed = False
    for path in resumes:
        try:
            print(f"PDF: {build(path, template, output_dir)}")
        except Exception as exc:
            print(f"FAILED {path}: {exc}", file=sys.stderr)
            failed = True
    return failed


def mtimes(files):
    return {file: file.stat().st_mtime for file in files}


def watch(resumes, template, output_dir):
    watched = [*resumes, template]
    last = None
    print("Watching for changes. Press Ctrl+C to stop.")
    while True:
        current = mtimes(watched)
        if current != last:
            build_all(resumes, template, output_dir)
            last = current
        time.sleep(0.5)


def main():
    args = parse_args()
    resumes = paths(args.resumes)
    if args.watch:
        try:
            watch(resumes, args.template, args.output_dir)
        except KeyboardInterrupt:
            print("\nStopped watching.")
            return
    sys.exit(build_all(resumes, args.template, args.output_dir))


if __name__ == "__main__":
    main()
